### PR TITLE
Add internal events feed

### PR DESCRIPTION
**Pull request by costinm**
Experimental 'events over XDS':

https://docs.google.com/document/d/1X7DmkP10x4nEcTGpikxmvLbBTTK59B0WL3liQzVy8Ms/edit#heading=h.xw1gqgyqs5b

Prototype and initial test impl for evaluating the XDS events. 

May be removed if better option is found.

### CODE REVIEW

**Comment by costinm** [comment]
(I'm doing some updates based on the feedback in the WIP PR - and adding tests)

File: pilot/pkg/proxy/envoy/v2/internalgen.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	// TODO: TypeURLReady - readiness events for endpoints, agent can propagate
+
+	// TypeURLNACK will receive messages of type DiscoveryRequest, containing
+	// the 'NACK' from envoy on rejected configs. Only ID is set in metadata.
+	// This includes all the info that envoy (client) provides.
+	TypeURLNACK = "istio.io/nack"
+
+)
+
+// InternalGen is a Generator for XDS status updates: connect, disconnect, nacks, acks
```
Comment (howardjohn) [inline pilot/pkg/proxy/envoy/v2/internalgen.go:39]: This doesn't really seem to be just a "generator". Maybe we should make the Generate api more generic and add some Push(...) method or similar? The core XDS will need it anyways when/if it moves to this model?

File: pilot/pkg/proxy/envoy/v2/internalgen.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+type InternalGen struct {
+	Server *DiscoveryServer
```
Comment (howardjohn) [inline pilot/pkg/proxy/envoy/v2/internalgen.go:41]: If we do the above comment and move the pushing logic out of here, we can avoid this circular dependency

File: pilot/pkg/proxy/envoy/v2/internalgen.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+	// TODO: track last N Nacks and connection events, with 'version' based on timestamp.
+	// On new connect, use version to send recent events since last update.
+}
+
+func (sg *InternalGen) OnConnect(node *core.Node) {
+	sg.startPush(TypeURLConnections, []*any.Any{util.MessageToAny(node)})
+}
+
+func (sg *InternalGen) OnDisconnect(node *core.Node) {
```
Comment (howardjohn) [inline pilot/pkg/proxy/envoy/v2/internalgen.go:51]: nit: I think our code should rely on istio objects more than envoy where possible. Would recommend taking a model.proxy here, then we can just call proxy.node to get the core.Node from it. This way if/when we add auto registration of XDS clients as endpoints we won't need to refactor

File: pilot/pkg/proxy/envoy/v2/internalgen.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+// Generate XDS responses about internal events:
+// - connection status
+// - NACKs
+//
+// We can also expose ACKS.
+func (sg *InternalGen) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
+	res := []*any.Any{}
+	switch w.TypeUrl {
+	case TypeURLConnections:
```
Comment (howardjohn) [inline pilot/pkg/proxy/envoy/v2/internalgen.go:96]: Why only TypeURLConnections and not the others?

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 5]
```diff
@@ -349,6 +361,10 @@ func (s *DiscoveryServer) handleCds(con *XdsConnection, discReq *xdsapi.Discover
 			errCode := codes.Code(discReq.ErrorDetail.Code)
 			adsLog.Warnf("ADS:CDS: ACK ERROR %s %s:%s", con.ConID, errCode.String(), discReq.ErrorDetail.GetMessage())
 			incrementXDSRejects(cdsReject, con.node.ID, errCode.String())
+			if s.InternalGen != nil {
```
Comment (howardjohn) [inline pilot/pkg/proxy/envoy/v2/ads.go:364]: Alternative to all the `if != nil` is to make `nil.OnNack` do nothing instead of panic. Not sure if that is more or less confusing though

**Comment by howardjohn** [comment]
```
grpcurl -d @ -insecure localhost:15012 envoy.service.discovery.v2.AggregatedDiscoveryService/StreamAggregatedResources <<EOM | jq '.resources[].id'
{
  "node": {"id": "sidecar~1.1.1.1~debug~cluster.local", "metadata": {"GENERATOR": "event"}},
  "typeUrl": "istio.io/connections"
}
EOM
```

To test locally with CLI - depends on https://github.com/istio/istio/pull/23964

@esnible

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 6]
```diff
@@ -111,6 +111,10 @@ type XdsConnection struct {
 		RDS string
 		LDS string
 	}
+
+	// Original node metadata, to avoid unmarshall/marshall. This is included
+	// in internal events.
+	meta *core.Node
```
[the remark this answers is not in the document]
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/ads.go:117]: node.Metadata should be sufficient. Later we'll probably add a proper proto with the info we really need ( labels, IP ). 

I do want to track the Istiod address as well.

[the remark this answers is not in the document]
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/ads.go:117]: Actually - Id, Locality, Cluster may be useful as well. 

I would keep it - we send the node pretty often from envoy.

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 7]
```diff
@@ -488,9 +510,13 @@ func (s *DiscoveryServer) initConnection(node *core.Node, con *XdsConnection) er
 	con.mu.Lock()
 	con.node = proxy
 	con.ConID = connectionID(node.Id)
+	con.meta = node
```
[the remark this answers is not in the document]
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/ads.go:513]: renamed to xdsNode

File: pilot/pkg/proxy/envoy/v2/internalgen.go (hunk 1 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/internalgen.go:39]: Maybe - but in separate PR ? 

The model is really request -> response + updates. 
Push is a bit hard to abstract - for example in this case we want the push to be sent directly, in other cases we want to debounce. The 'generator' doesn't really care about push - is a different code path. 

In any case - I'm sure we'll iterate and improve the abstractions as we start using it more.

File: pilot/pkg/proxy/envoy/v2/internalgen.go (hunk 2 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/internalgen.go:41]: Right - the methods should probably by DiscoveryServer methods, on in InternalGen. 

Can we refactor in separate PR ?

File: pilot/pkg/proxy/envoy/v2/internalgen.go (hunk 3 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/internalgen.go:51]: That's a bit controversial. 
model.Proxy is not a proto - so we can't really use it. We could define a proto - which is one of the possible long term paths. 

But at the same time this is XDS - and maybe the 'connect'/disconnect should actually be XDS top level objects instead of istio specific. There is nothing special about Istio in this event. 

That was  the main reason I used 'node' - any XDS server can implement this without taking a dependency on Istio. 

I was thinking even long term if we define the Istio specific object - we can still support the type URL based on envoy object. And ideally a top level envoy type could be defined that would be used instead of Node.

File: pilot/pkg/proxy/envoy/v2/internalgen.go (hunk 4 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/internalgen.go:96]: Nacks - we don't store/cache, it would be used as: register listener, make config change, watch for nacks. We could have a LRU with recent nacks if we want.

Disconnect - I'm planning to remove it. For experimenting it seemed useful - but I'll look for more  feedback. And there is nothing to return if we treat it as a REST ( I think you had a comment about this in the 'WIP' PR)

File: pilot/pkg/proxy/envoy/v2/internalgen.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/proxy/envoy/v2/internalgen.go:51]: You can still make OnDisconnect accept a model.Proxy though but send only the core.Node. The reason I suggested this is because its needed for the auto-registration. Agreed actually sending XDS/UDPA objects on the wire is probably a good idea, at least for now

File: pilot/pkg/proxy/envoy/v2/internalgen.go (hunk 2 above)
Comment (howardjohn) [reply pilot/pkg/proxy/envoy/v2/internalgen.go:41]: Sure

File: pilot/pkg/proxy/envoy/v2/ads.go (hunk 5 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/ads.go:364]: I have a bit of a panic attack when I read things like nil.OnNack(). Many years of java and c.

Comment (howardjohn) [reply pilot/pkg/proxy/envoy/v2/ads.go:364]: Yeah thats probably for the best, probably leads to more confusion than it is worth

**Comment by costinm** [comment]
> ```
> grpcurl -d @ -insecure localhost:15012 envoy.service.discovery.v2.AggregatedDiscoveryService/StreamAggregatedResources <<EOM | jq '.resources[].id'
> {
>   "node": {"id": "sidecar~1.1.1.1~debug~cluster.local", "metadata": {"GENERATOR": "event"}},
>   "typeUrl": "istio.io/connections"
> }
> EOM
> ```
> 
> To test locally with CLI - depends on #23964
> 
> @esnible

Cool - in the main WIP PR I had some fixes for pilot-tool - but using grpcurl instead seems better !

**Comment by esnible** [comment]
@howardjohn I tried your example.

https://github.com/istio/istio/pull/23964 merged so I tested with _grpcurl_.

The output is empty.  Everything is working, because if I leave off the `node` in the query I get `Message: missing node ID`.  I am unclear what to supply as a node ID to get output.

I am surprised that the port is 15012 because that port is not exposed by the pod and the _istiod_ Service suggests 15012 is DNS.

```
  - name: grpc-xds
    port: 15010
    protocol: TCP
    targetPort: 15010
  - name: https-dns
    port: 15012
    protocol: TCP
    targetPort: 15012
```

File: pkg/adsc/adsc.go [hunk 8]
```diff
@@ -782,7 +824,22 @@ func (a *ADSC) Wait(to time.Duration, updates ...string) ([]string, error) {
...
+			// legacy names, still used in tests.
+			switch t {
+			case ListenerType:
+				delete(want, "lds")
+			case v2.ClusterTypeV3:
+				delete(want, "cds")
+			case v2.EndpointTypeV3:
+				delete(want, "eds")
+			case routeType:
+				delete(want, "rds")
```
Comment (nrjpoddar) [inline pkg/adsc/adsc.go:840]: @costinm  can you explain why is this change needed?

File: pilot/pkg/proxy/envoy/v2/internalgen.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+	// Make sure we include the ID - the DR may not include metadata
+	dr.Node.Id = node.ID
+	sg.startPush(TypeURLNACK, []*any.Any{util.MessageToAny(dr)})
+}
+
+// startPush is similar with DiscoveryServer.startPush() - but called directly,
+// since status discovery is not driven by config change events.
+// We also want connection events to be dispatched as soon as possible,
+// they may be consumed by other instances of Istiod to update internal state.
+func (sg *InternalGen) startPush(typeURL string, data []*any.Any) {
```
Comment (nrjpoddar) [inline pilot/pkg/proxy/envoy/v2/internalgen.go:64]: Should there be unit tests for these methods?

File: pilot/pkg/proxy/envoy/v2/internalgen.go (hunk 3 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/internalgen.go]: Got it, sounds good - will make the change.

File: pkg/adsc/adsc.go (hunk 8 above)
Comment (costinm) [reply pkg/adsc/adsc.go:840]: Sync up with the V3 change and make the code less dependent of the old endpoints. 

ADSC is a client for XDS/ADS - was initially used for testing, but I'm planning to use it in agent for pulling ServiceEntry from Istiod so it can responds to DNS without a RTT. 
It needs a bit of cleanup...

File: pilot/pkg/proxy/envoy/v2/internalgen.go (hunk 9 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/internalgen.go:63]: There is a unit test in the PR - it checks that an initial response is sent and that on a new connection there is a push with the event.

Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/internalgen.go:63]: And more tests will be added as it gets used - for now it's experimental, we need to find out the right interface based on the various use cases. John is using this for the config-less multi cluster, it may be used for debug - and I'm planning to use it for some optimizations on discovery propagation latency.

**Comment by costinm** [comment]
> @howardjohn I tried your example.
> 
> #23964 merged so I tested with _grpcurl_.
> 
> The output is empty. Everything is working, because if I leave off the `node` in the query I get `Message: missing node ID`. I am unclear what to supply as a node ID to get output.
> 
> I am surprised that the port is 15012 because that port is not exposed by the pod and the _istiod_ Service suggests 15012 is DNS.
> 
> ```
>   - name: grpc-xds
>     port: 15010
>     protocol: TCP
>     targetPort: 15010
>   - name: https-dns
>     port: 15012
>     protocol: TCP
>     targetPort: 15012
> ```

15012 is the https port

**Comment by costinm** [comment]
The command I use manually:
```
echo '{"node": {"id": "sidecar~1.1.1.1~debug~cluster.local", "metadata": {"GENERATOR": "event", "foo":"bar"}},"typeUrl": "istio.io/connections"}' | grpcurl -d @ -insecure localhost:15012 envoy.service.discovery.v2.AggregatedDiscoveryService/StreamAggregatedResources
```

That actually closes the connection, if you want to see events:

```
grpcurl -d @ -insecure localhost:15012 envoy.service.discovery.v2.AggregatedDiscoveryService/StreamAggregatedResources

and  type in the input:
{"node": {"id": "sidecar~1.1.1.1~debug~cluster.local", "metadata": {"GENERATOR": "event", "foo":"bar"}},"typeUrl": "istio.io/connections"}
```
